packages/vite/src/node/plugins/splitVendorChunk.ts

```diff
@@ -0,0 +1,95 @@
+import type { Plugin } from '../plugin'
+import type { OutputOptions, GetManualChunk, GetModuleInfo } from 'rollup'
+import { isCSSRequest } from './css'
+
+// Use splitVendorChunkPlugin() to get the same manualChunks strategy as Vite 2.7
+// We don't recommend using this strategy as a general solution moving forward
+
+// splitVendorChunk is a simple index/vendor strategy that was used in Vite
+// until v2.8. It is exposed to let people continue to use it in case it was
+// working well for their setups.
+// The cache needs to be reset on buildStart for watch mode to work correctly
+// Don't use this manualChunks strategy for ssr, lib mode, and 'umd' or 'iife'
+
+export class SplitVendorChunkCache {
+  cache: Map<string, boolean>
+  constructor() {
+    this.cache = new Map<string, boolean>()
+  }
+  reset() {
+    this.cache = new Map<string, boolean>()
+  }
+}
+
+export function splitVendorChunk({
+  cache = new SplitVendorChunkCache()
+}): GetManualChunk {
+  return (id, { getModuleInfo }) => {
+    if (
+      id.includes('node_modules') &&
+      !isCSSRequest(id) &&
+      staticImportedByEntry(id, getModuleInfo, cache.cache)
+    ) {
+      return 'vendor'
+    }
+  }
+}
+
+function staticImportedByEntry(
+  id: string,
+  getModuleInfo: GetModuleInfo,
+  cache: Map<string, boolean>,
+  importStack: string[] = []
+): boolean {
+  if (cache.has(id)) {
+    return cache.get(id) as boolean
+  }
+  if (importStack.includes(id)) {
+    // circular deps!
+    cache.set(id, false)
+    return false
+  }
+  const mod = getModuleInfo(id)
+  if (!mod) {
+    cache.set(id, false)
+    return false
+  }
+
+  if (mod.isEntry) {
+    cache.set(id, true)
+    return true
+  }
+  const someImporterIs = mod.importers.some((importer) =>
+    staticImportedByEntry(
+      importer,
+      getModuleInfo,
+      cache,
+      importStack.concat(id)
+    )
+  )
+  cache.set(id, someImporterIs)
+  return someImporterIs
+}
+
+export function splitVendorChunkPlugin(): Plugin {
+  const cache = new SplitVendorChunkCache()
+  return {
+    name: 'vite:split-vendor-chunk',
+    config(config) {
+      const build = config.build ?? {}
+      const format = (build.rollupOptions?.output as OutputOptions)?.format
+      if (!build.ssr && !build.lib && format !== 'umd' && format !== 'iife') {
```

Comment:
We should abstract this away, so chunk plugins don't all have to repeat and maintain it.

I suggest we move this check into Vite core, like below:
```ts
// in `resolveConfig` maybe or where Rollup `output` options are generated...

if (build.ssr || build.lib || format === 'umd' || format === 'iife') {
  build.rollupOptions.manualChunks = undefined
}
```

Reply:
I don't think we want to keep this complexity in Core. There is also a high chance we will not offer other chunking strategies. And in this particular case, are we sure that every possible manualChunks value is invalid for these formats and options?

Reply:
> There is also a high chance we will not offer other chunking strategies.

I'm assuming all chunk plugins (not just the ones Vite offers) will want this logic.



> are we sure that every possible manualChunks value is invalid for these formats and options?

Idk, maybe @yyx990803 has some insight there.

Reply:
If we want something like this, we may expose something like `isManualChunksAllowed` function later? I'm not convinced we should stop the user from setting `manualChunks` without a way out in all these cases. Til now it was "under these conditions, slit vendor chunk shouldn't be applied", and you are proposing to change that to "under these conditions, manualChunks is disallowed" that isn't the same.